jujugui/static/gui/src/app/state/test-state.js

```diff
@@ -744,12 +706,88 @@ describe('State', () => {
 
       badURLS.forEach(test => {
         assert.deepEqual(
-          state.buildState(test.path),
+          state.generateState(test.path),
           {error: test.error, state: test.state},
           `${test.path} did not properly generate the state object: ` +
           JSON.stringify(test.state));
       });
     });
+  });
+
+  describe('State.dispatch()', () => {
+    it('passes the current location to generateState', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123',
+        seriesList:  ['precise', 'trusty', 'xenial'],
+        location: {href: 'foo'}
+      });
+      console.log(state.generateState);
+      const stub = sinon.stub(
+        state, 'generateState', () => ({ error: null, state: {}}));
+      state.dispatch();
+      assert.equal(stub.callCount, 1);
+      assert.deepEqual(stub.args[0], ['foo']);
+    });
+
+    it('updates the _appStateHistory with the new state', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123',
+        seriesList:  ['precise', 'trusty', 'xenial'],
+        location: {href: 'foo'}
+      });
+      sinon.stub(state,
+        'generateState', () => ({ error: null, state: {new: 'state'}}));
+      state.dispatch();
+      assert.deepEqual(state._appStateHistory, [{new: 'state'}]);
+    });
+  });
+
+  describe('State.generatePath()', () => {
```

Comment:
This uses identical data to the url generation so we only have to add/modify the test data in one place and it'll be tested on both directions of the state system.